wagtail/wagtailadmin/navigation.py

```diff
@@ -0,0 +1,103 @@
+from __future__ import absolute_import, unicode_literals
+
+from django.db.models import Q
+
+from wagtail.wagtailcore.models import Page
+
+
+def get_navigation_menu_items(user):
+    # Get all pages that the user has direct add/edit/publish/lock permission on
+    if user.is_superuser:
+        # superuser has implicit permission on the root node
+        pages_with_direct_permission = Page.objects.filter(depth=1)
+    else:
+        pages_with_direct_permission = Page.objects.filter(
+            group_permissions__group__in=user.groups.all(),
+            group_permissions__permission_type__in=['add', 'edit', 'publish', 'lock']
+        )
+
+    if not(pages_with_direct_permission):
+        return []
+
+    # Find the closest common ancestor of the pages the user has permission for;
+    # this (or its children) will be the root menu level for this user.
+    cca_path = pages_with_direct_permission[0].path
```

Comment:
Probably, yep :-) It would involve an extra query though (because we don't use the resulting Page instance here, only the path).